### ISSUE TITLE
[need-help] Align Cornice with Pyramid ACL pattern

### ISSUE THREAD

**Issue by leplatrem** [Cornices/cornice#452]
Cornice comes in the way of several developers when it comes to dynamic ACLs, especially for resources.

The following is a very elegant pattern [proposed](https://github.com/Cornices/cornice/issues/120#issuecomment-284136414) by @DaineDanielson

```py
@resource(collection_path='/pages', path='/pages/{id}')
class PageResource:
    def __init__(self, request):
        self.id = uuid4()
        self.creator_id = request.user_id

    def __acl__(self):
        return [(Allow, Authenticated, 'add'),
                (Allow, self.creator_id, 'read')]

    @view(permission='add')
    def collection_post(self):
        self.request.response.status_code = 201
        return self.id
     
    @view(permission='read')
    def get(self):
        # ... retrieve page here ...
        return page
```

I'm not super available for this, so if someone is interested to lead that work, I would be happy to review, merge contributions and give admin privileges on the repo :)

**Comment by wjehenddher** [comment]
It seems like using the `factory` argument, the scenario above can be achieved almost out of the box. 

```python
@resource(collection_path='/pages', path='/pages/{id}',

    factory=__name__ + ".PageResource"

)
class PageResource:
    def __init__(
        self, 
        request,
        # to prevent mismatched signature from cornice.service.Service.decorate_view.wrapper 
        #        ob = args['klass'](**params)  
        *args,
        **kwargs
    ):
        self.id = uuid4()
        self.creator_id = request.user_id

    def __acl__(self):
        return [(Allow, Authenticated, 'add'),
                (Allow, self.creator_id, 'read')]

    @view(permission='add')
    def collection_post(self):
        self.request.response.status_code = 201
        return self.id
     
    @view(permission='read')
    def get(self):
        # ... retrieve page here ...
        return page
```

Should the new implementation automatically wire `factory` to the `@resource` decorated class if no `klass` is explicitly declared?

Thoughts?

**Comment by wjehenddher** [comment]
@leplatrem et al,

I pinged Pyramid IRC and got very good info from **raydeo**, one of the moderators.

In Pyramid theory, `PageResource.__acl__()` should be agnostic of `PageResource.id` and hence agnostic of the actual `request`. Instead, `PageResource.__acl__` should return the full list of ACEs ignoring current user and/or current request.

It's worth to note that the approach we are pursuing, *allowing `PageResource.__acl__` change dynamically based on `request` (and/or `user`)*, would *philosophically* contradict the Pyramid semantics of `PageResource.__acl__`. 

From the practical point of view though, it will *arguably* be more efficient because the creation of `PageResource.__acl__` based on current user would be more efficient whenever ACEs are based on individual resources and their own users.

Consider this case (my target scenario): 

Each `PageResource` has multiple `Users` that are editors. But each `User` has access to zero or more `PageResources`

```
# 1 - user:1 creates a page (page:1) and becomes the creator
> POST /pages/
< 200, Location=/pages/1

# 2 - user:2 attempts to edit page:1 but fails with forbidden
> PUT /pages/1
< 403

# 3 - user:1 grants editor rights to user:2
> PUT /pages/1/editors/user2
< 200

# 4 - User 2 edits page:1 successfully
> PUT /pages/1
< 200
```

For this use case, `PageResource` could look like this:

```
def __acl__(self):

    return [
        (Allow, Authenticated, 'add'),
        (Allow, self.creator_id, 'read'),
        (Allow, 
           self.request.authenticated_userid, 
           self.local_roles_for_user(self.request.authenticated_userid)
    ]
         
def local_roles_for_user(self, userid):
    """ Returns the list of roles of a given user has considering to the current request/resource being accessed """
    return self.dbsession.getUserPermissionForResource(self.resource.id) # ("edit", "view")
```

Would you agree that this is the scenario we want to achieve?

**raydeo** pointed out that a custom `AuthorizationPolicy` could do the trick for what I am looking for: https://pypi.org/project/pyramid_localroles/

**Comment by wjehenddher** [comment]
@leplatrem 

I wonder if we have 2 issues at hand:

1. How `@resource` (and internal `service`/`_UnboundView`) work with pyramid `acl` considering that `PageResource` could automatically be the `factory`
2. Dynamic `acl` based on resource (aka request and user aware acl)

**Comment by DaineDanielson** [comment]
First, thank you all for the vigorous discussion! If I weren't 110% full-time building my company's stack (Cornice-based back-end, naturally), I would be fully engaged here. Keeping that in mind, you'll forgive me if I'm very succinct in targeting this reply.

@wjehenddher, I'm responding to the comment that
> In Pyramid theory, PageResource.__acl__() should be agnostic of PageResource.id and hence agnostic of the actual request. Instead, PageResource.__acl__ should return the full list of ACEs ignoring current user and/or current request.

I'll push back on one sub-point here from a Pyramid-theoretic perspective: while I agree it may be non-purist for `__acl__` to depend on the incoming request, I think it _is_ strictly Pyramidic for `__acl__` to depend on `PageResource.id`, or, more generally, for `__acl__` to depend on the specific resource instance being requested. Looking back to `PageResource.__acl__` [in this official example](https://docs.pylonsproject.org/projects/pyramid/en/latest/tutorials/wiki2/authorization.html#add-resources-and-acls), you can see `__acl__` depends dynamically on the specific `Page` instance `self.page`, which is functionally equivalent to depending dynamically on `PageResource.id` in my pattern, as shared by [OP](https://github.com/Cornices/cornice/issues/452#issue-248023099) @leplatrem.

**Comment by wjehenddher** [comment]
Thank you @DaineDanielson. We are on the same page: for practical purposes, the scope of this issue is to allow `__acl__` be dynamic and aware of the current Resource.

That takes care of the second bullet in https://github.com/Cornices/cornice/issues/452#issuecomment-321039850.

For the first bullet:
> How `@resource` (and internal `service/_UnboundView`) would work with pyramid considering that `PageResource` could automatically be the factory and hence be able to return dynamic acl

Should we automatically make the decorated class the factory?
The idea is to replace default root factory (`request.context`) with the decorated class.

(something along these lines)

```python
# 'factory' is not declared, use decorated class
@resource(collection_path='/pages', path='/pages/{id}')
class PageResource:
...

# cornice/resource.py

def resource(...):
     ...
     # No 'factory' declared, make klass the factory
     if "factory" not in service_args:
         service_args["factory"] = klass
     ...
     service = services[service_name] = Service(name=service_name, ...
```
Of course,  every decorated class would need to define its` __init__(request, *args, **kwargs)` so there is no mismatched signature as I pointed out in https://github.com/Cornices/cornice/issues/452#issuecomment-320796113

So you know, our usage of cornice and pyramid is vanilla. Hence, we don't combine/use many of the available view/factory/service arguments. Changes like these, makes me worry about breaking existing functionality.

**Comment by wjehenddher** [comment]
@leplatrem @DaineDanielson 

Any thoughts about my proposal for automatically wiring `factory` argument to `klass` when `@resource` is resolved (iff no `factory` argument was explicitly declared)?

In our baseline, I am now explicitly declaring `factory` and adding `__acl__` to our `@resource` decorated classes.

**Comment by wjehenddher** [comment]
@leplatrem, @DaineDanielson 

I'd like to avoid pursuing an ill founded option. Especially, given that the time I can allocate to this is limited.

Thoughts about the auto-wiring of `factory`?

**Comment by leplatrem** [comment]
Hi everyone, I'm sorry for my late response, this got burried (summer time etc.).

If I understand correctly the only drawback of your proposal is that if `factory` is not specified then the class should implement a `__init__()`. 

Honestly, I find it acceptable, and I don't mind bumping a major version if that's justified.

**Comment by wjehenddher** [comment]
Thank you @leplatrem

I think the `__init__(...)` is perhaps an *unfortunate side effect* of specifying `factory` argument in the `@resource()` decoration. I am sure it can be addressed reasonably well (as you pointed out, even if it needs bumping up a major version). No worries there.

At any rate, here is where I stand now:

Without any changes to cornice (v1.0.0 for us), I have successfully achieved dynamic `__acl__` at the `@resource` level by using a *super abstract class* `BaseResource`.

`BaseResource` implements `__acl__` and expects each subclass to return the resource being affected, based on subclass' `request.matched_route.pattern` .  I.e. `BaseResource` takes into consideration the current `request.authenticated_userid` and it returns the `ACE`s the authenticated user is allowed on the resource. 

My current implementation looks like this (more or less):

```python

class BaseResource(object):

    def __init__(self, request):
        self.request = request

    def get_resource(self):
        """
        Return the resource based on request's matched route.
        Subclasses are required to override this method
        Ideally, abc should be used (https://docs.python.org/2/library/abc.html)
        """
        return None

    def __acl__(self):
        acl = [ # hardcoded ACEs are be declared here
             (Allow, self.request.unauthenticated_userid, ()), # unauthenticated user can't do anything
             (Allow, self.request.authenticated_userid, ('add', )), # authenticated user can add new pages
        ]

        # Ask subclass to return the affected resource (based on matched route or whatever else)
        resource = self.get_resource()

        authuser_aces = self.request.db.get_authenticated_userid_permissions_on_resource(resource, self.request.authenticated_userid)
        # authuser_aces would look like [ (Allow, authenticated_userid, ('read',)) ] if user created page.id == id

        return acl.extend(authuser_aces)


@resource(collection_path='/pages', path='/pages/{id}',
    factory=__name__ + ".PageResource" # boilerplate (1)
)
class PageResource(BaseResource):

    def __init__(
        self, 
        request,
        # boilerplate (2)
        # to prevent mismatched signature from cornice.service.Service.decorate_view.wrapper 
        #        ob = args['klass'](**params)  
        # request=..., context=... are being passed to __init__(...)
        *args,
        **kwargs
    ):
        BaseResource.__init__(self, request)

    def get_resource(self):
        if self.request.matched_route.pattern == '/pages/{id}':
            page_id = self.request.matchdict["id"]
            return self.request.db.get_page_by_id(page_id)
        else:
            return None

    @view(permission='add')
    def collection_post(self):
        # add new page ...
        ...

    @view(permission='read')
    def get(self):
        # retrieve page ...
        return page
```
 
In summary, as a Cornice (v1.0.0) user, I just need to add that bit of boilerplate code to each of my `@resource`s and voila ... dynamic `__acl__` 

IMHO, it seems like the only change we would need to make to cornice is to automatically wire `factory` to the decorated class (1) and fix the `__init__()` (2). Again, just with the goal of removing the boilerplate code.

At the end, Cornice integrators would **still** have the freedom to wire their own `factory` (by specifying one in the `@resource` decorator, or via pyramid's `Configurator` `root_factory`), or use a custom class hierarchy to consolidate redundant ACL code, or whatever else they want.

What do you all think?

**Comment by wjehenddher** [comment]
Hello @leplatrem et al

As I do the *auto-wire* of `klass` as `factory`, test-cases break. Specifically those that declared an `acl`. They are breaking because I am "discarding" the declared `acl` whenever *auto-wire* happens.

```
ERROR: test_acl_support_authenticated_allowed_thing_get (tests.test_imperative_resource.TestResource)

ERROR: test_acl_support_authenticated_allowed_thing_get (tests.test_resource.TestResource)

ERROR: test_collection_acl_can_be_different (tests.test_resource.TestResource)
```

However, it's my understanding that param `acl` was deprecated though.
https://github.com/Cornices/cornice/blob/master/cornice/service.py#L284

Should I worry about keeping the existing test-cases dealing with `acl` backward compatible? 
(even though bumping the major version number was justified).

I could just fix the broken tests so they check the new conditions but they would no longer be asserting the behavior in previous version.

I'll pushing the PR soon to ease matters.

**Comment by leplatrem** [comment]
> Should I worry about keeping the existing test-cases dealing with acl backward compatible?

If the `acl` parameter is deprecated, then nope, we can get rid of it, since we'll probably bump a major with your changes

**Comment by wjehenddher** [comment]
Hi @lepraten

I think the only thing left to deal with is the `__init__(self, request, context=None)` issue; assuming we really want to address it.

I ran into a problems trying to remove it. See https://github.com/Cornices/cornice/pull/456#discussion_r138968777

I think I need understand better the intricacies of `cornice.service.decorate_view()` and its `args` and `route_args`. 

If you can shade some light, I'd appreciate it.

**Comment by wjehenddher** [comment]
Hello @leplatrem and @tarekziade 

My attempts to prevent changes to the `__init__` method were not fruitful.

Nonetheless, I made changes to documentation here: 55d7e947c69c4

Thoughts and comments are welcome!

**Comment by leplatrem** [comment]
Fixed by #456
